tutorials/getting_started/laziness.md

```diff
@@ -0,0 +1,233 @@
+            # instance = ...
+            instances.append(instance)
+        return instances
+```
+
+but should instead do
+
+```python
+    def _read(self, file_path: str) -> Iterable[Instance]:
+        """Do write your _read function this way"""
+        # logic to iterate over file
+            # some kind of for loop
+            # instance = ...
+            yield instance
+```
+
+In the `lazy=False` case, `DatasetReader.read()` will
+materialize your instances to a list before returning them anyway.
+
+## Laziness in `experiment.json`
+
+If you implemented your dataset reader as specified above,
+laziness just requires adding `"lazy": true` to your
+dataset reader config:
+
+```json
+  "dataset_reader": {
+    "type": "snli",
+    "lazy": true,
+    "token_indexers": {
+      "tokens": {
+        "type": "single_id",
+        "lowercase_tokens": true
+      }
+    }
+  },
+```
+
+## Laziness in `DataIterator`
+
+AllenNLP uses a `DataIterator` abstraction to iterate over datasets
+using configurable batching, shuffling, and so on.  The included
+dataset readers all work with lazy datasets, but they have extra options
+that you might want to specify in that case. In particular, we'll
+look at the
+[`BasicIterator`](https://github.com/allenai/allennlp/blob/master/allennlp/data/iterators/basic_iterator.py):
+
+```python
+    def __init__(self,
+                 batch_size: int = 32,
+                 instances_per_epoch: int = None,
+                 max_instances_in_memory: int = None) -> None:
+        self._batch_size = batch_size
+        self._instances_per_epoch = instances_per_epoch
+        self._max_instances_in_memory = max_instances_in_memory
+```
+
+If you use the default parameters and call this iterator on a lazy dataset,
+it will lazily read 32 instances at a time, package those instances into a
+`Batch`, and yield them one by one. If you specify `shuffle=True` when you call
+the iterator, it will shuffle the 32 instances within each batch, but will always
+serve batches in the same order.
+
+If you have a sense for how many instances can fit in memory, you can specify
+the `max_instances_in_memory` parameter. For instance, if you were to set that
+to 3200, the iterator would load 3200 instances at a time into memory,
+potentially shuffle them, and then create 100 batches of size 32 and yield
+them up before loading the next 3200 instances.
+
+This is especially useful if you are using the `BucketIterator`,
+which can order your instances by e.g. sentence length. If you don't specify
+`max_instances_in_memory`, you'd just get each batch of 32 ordered by sentence length,
+which isn't particularly helpful.
+
+The other option here is `instances_per_epoch`. By default, each epoch is a single
+pass through the dataset. However, if you had millions of instances,
+you might want each epoch to consist of a fraction of the dataset,
+in which case you can specify the value here. But most likely you don't need
```

Comment:
Specifying this will also result in nicer tqdm logging.

Reply:
the way things are designed right now that's a problem unless your train / validation / test sets are the same size.

i.e. if you have 1000 train examples / 200 validation examples, and if you specify `instances_per_epoch: 1000`, then when you iterate over the validation set it will make 5 passes over the data

this is why I was earlier suggesting making the dataset parts of the configuration slightly richer -- it seems likely you'd want to specify different values for things like this for different datasets

Reply:
or another alternative is just to count the instances the first pass through the dataset and remember the number for future passes

Reply:
Oh, yeah, you're right, I forgot about that.  Mandar said he was going to submit a PR to allow for different train and validation iterators, which would solve this problem.